File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/AdapterViewSelectionObservable.java
package com.jakewharton.rxbinding2.widget;

import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static android.widget.AdapterView.INVALID_POSITION;
import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class AdapterViewSelectionObservable
    extends InitialValueObservable<AdapterViewSelectionEvent> {
  private final AdapterView<?> view;

  AdapterViewSelectionObservable(AdapterView<?> view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super AdapterViewSelectionEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    view.setOnItemSelectedListener(listener);
    observer.onSubscribe(listener);
  }

  @Override protected AdapterViewSelectionEvent getInitialValue() {
    int selectedPosition = view.getSelectedItemPosition();
    if (selectedPosition == INVALID_POSITION) {
      return AdapterViewNothingSelectionEvent.create(view);
    } else {
      View selectedView = view.getSelectedView();
      long selectedId = view.getSelectedItemId();
      return AdapterViewItemSelectionEvent.create(view, selectedView, selectedPosition, selectedId);
    }
  }

  static final class Listener extends MainThreadDisposable implements OnItemSelectedListener {
    private final AdapterView<?> view;
    private final Observer<? super AdapterViewSelectionEvent> observer;

    Listener(AdapterView<?> view, Observer<? super AdapterViewSelectionEvent> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
      if (!isDisposed()) {
        observer.onNext(AdapterViewItemSelectionEvent.create(parent, view, position, id));
      }
    }

    @Override public void onNothingSelected(AdapterView<?> parent) {
      if (!isDisposed()) {
        observer.onNext(AdapterViewNothingSelectionEvent.create(parent));
      }
    }

    @Override protected void onDispose() {
      view.setOnItemSelectedListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding3/widget/AdapterViewSelectionObservable.kt
@file:JvmName("RxAdapterView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.widget

import android.view.View
import android.widget.Adapter
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import com.jakewharton.rxbinding3.InitialValueObservable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import android.widget.AdapterView.INVALID_POSITION
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of selection events for `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun <T : Adapter> AdapterView<T>.selectionEvents(): InitialValueObservable<AdapterViewSelectionEvent> {
  return AdapterViewSelectionObservable(this)
}

private class AdapterViewSelectionObservable(
  private val view: AdapterView<*>
) : InitialValueObservable<AdapterViewSelectionEvent>() {

  override fun subscribeListener(observer: Observer<in AdapterViewSelectionEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    view.onItemSelectedListener = listener
    observer.onSubscribe(listener)
  }

  override val initialValue: AdapterViewSelectionEvent
    get() {
      val selectedPosition = view.selectedItemPosition
      return if (selectedPosition == INVALID_POSITION) {
        AdapterViewNothingSelectionEvent(view)
      } else {
        val selectedView = view.selectedView
        val selectedId = view.selectedItemId
        AdapterViewItemSelectionEvent(view, selectedView, selectedPosition, selectedId)
      }
    }

  private class Listener(
    private val view: AdapterView<*>,
    private val observer: Observer<in AdapterViewSelectionEvent>
  ) : MainThreadDisposable(), OnItemSelectedListener {

    override fun onItemSelected(parent: AdapterView<*>, view: View?, position: Int, id: Long) {
      if (!isDisposed) {
        observer.onNext(AdapterViewItemSelectionEvent(parent, view, position, id))
      }
    }

    override fun onNothingSelected(parent: AdapterView<*>) {
      if (!isDisposed) {
        observer.onNext(AdapterViewNothingSelectionEvent(parent))
      }
    }

    override fun onDispose() {
      view.onItemSelectedListener = null
    }
  }
}
